esp-wifi/src/wifi/event.rs

```diff
@@ -0,0 +1,325 @@
+use alloc::boxed::Box;
+use core::cell::RefCell;
+
+use critical_section::Mutex;
+
+use super::WifiEvent;
+
+pub(crate) mod sealed {
+    use super::*;
+
+    pub trait Event {
+        /// Get the static reference to the handler for this event.
+        fn get_handler() -> &'static Mutex<RefCell<Option<Box<Handler<Self>>>>>;
+        /// # Safety
+        /// `ptr` must be a valid for casting to this event's inner event data.
+        unsafe fn from_raw_event_data(ptr: *mut crate::binary::c_types::c_void) -> Self;
+    }
+}
+/// The type of handlers of events.
+pub type Handler<T> = dyn FnMut(&T) + Sync + Send;
+
+fn default_handler<Event: 'static>() -> Box<Handler<Event>> {
+    fn drop_ref<T>(_: &T) {}
+    Box::new(drop_ref)
+}
+
+/// Extension trait for setting handlers for an event.
+///
+/// Register a new event handler like:
+/// ```
+/// # use esp_wifi::wifi::event::{self, *};
+/// # fn new_handler(_: &ApStaconnected) {}
+/// event::ApStaconnected::update_handler(|prev, event| {
+///     prev(event);
+///     new_handler(event);
+/// })
+/// ```
+// Implemented like this instead of free functions because the syntax would be
+// ```
+// event::update_handler::<event::ApStaconnected, _>(...)
+// ```
+pub trait EventExt: sealed::Event + Sized + 'static {
+    /// Get the handler for this event, replacing it with the default handler.
+    fn take_handler() -> Box<Handler<Self>> {
+        critical_section::with(|cs| {
+            Self::get_handler()
+                .borrow_ref_mut(cs)
+                .take()
+                .unwrap_or(default_handler::<Self>())
```

Comment:
This will allocate a new handler on each call. Despite the empty function, `drop_ref` is pointer-sized.

Reply:
I don't think so? `drop_ref` should be a zero sized function item. The pointer and vtable should fit in the reference. When I checked godbolt there was no assembly that allocated.